### PR TITLE
[USE MERGE COMMIT] LLVM Subtree update, September 2022

### PR DESCRIPTION
**Pull request by krzysz00**
In terms of our internal structure, this is a relatively un-disruptive update.

Note, **MIGraphX Breaking Change** due to changes to the C API headers upstream that removed `Registration.h`.

## Key changes
- Bufferization-related changes:
  - Now, the one-shot bufferizer no longer allows you to have a `bufferization.to_memref` in your input. Therefore, as a temporary hack, define `rock.tensor_conv2d` to represent convolutions in tensor land
  - Similarly, make `rock.transform` agnostic to whether its operand is a tensor or a memref
  - And finally, have `rock.tensor_untransform_cast` that lets us take the "result" of a `rock.tensor_conv2d` and interpret it as the raw, untransformed buffer so that the code after the `tosa.conv2d` doesn't complain.
  - This means that we now have a bufferization implementation registration hook on the rock dialect, which means that we have to call the function to register that hook. This means an extra API call to `registerRocMLIRDialects(registry)` in a few places, including everywhere that uses the C API
- Fixed the need to disable the `static_assert` in `OpDefinition.h` through using some templates in the generalized bufferization interface. Closes https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/595
- Moved all our passes to the autogenerated constructor infrastructure, eliminating the need for a `PassDetail.h` and getting out of the way of deprecation warnings. This has the advantage that our passes that take options now have a `BlahPassOptions` struct.
  - As a consequence, oved the options for AffixTuningParams to actual options
 - Since the pass boilerplate for a class was now a specifically `#include`d thing, made the Tosa partitioner **less** general: now it always takes a list of options.
 - Dodged a bunch of deprecation warnings related to `Option`, causing a lot of noise
 - The `operand_segment_sizes` attribute is now a DenseI32ArrayAttr, requiring mechanical changes

## Future work
- Clean up our build to include less of upstream's stuff and expose a us-specific registratino function https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/661
- Futz with any `DenseElementAttr`s we have running around our dialect to see if they can be `Dense[Blah]ArrayAttr`s https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/662
- Move our removal of the scoped context upstream https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/659
- The HIP headers moved in ROCm 5.2, account for that https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/658
- Define a `amdgpu.dpp` op to not have `gpu.warp_swizzle` running around https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/654
- Upstream the more generalized bufferization logic for calls and the requisite changes to the CallOpInterface https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/663
- Upstream Tosa partitioning https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/664
- Upstream `async.launch` https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/537
- Revisit https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/310, since I could drop the offending bit without consequence (*and* it still referred to `MLIR_MIOPEN`)
- Lean on the compiler team to add software BF16 support like what got added for X86 so that we can stop having `SoftwareBF16Pass`

### CODE REVIEW

**Comment by krzysz00** [comment]
Passing PR CI [xdlops job 3](http://rocmhead.amd.com:8080/blue/organizations/jenkins/MLIR%2Fmlir/detail/PR-779/3/pipeline) and [non-xdlops job 3](http://ml-ci.amd.com:21096/blue/organizations/jenkins/MLIR%2Fmlir/detail/PR-779/3/pipeline)

**Comment by krzysz00** [comment]
Current attempt at a nightly run. I claim that, since this PR doesn't touch our kernel generation logic, we're good to merge if we get a nightly pass and don't need a weekly http://rocmhead.amd.com:8080/job/MLIR/job/mlir/job/PR-779/5/

**Comment by krzysz00** [comment]
Nighly pass! http://rocmhead.amd.com:8080/blue/organizations/jenkins/MLIR%2Fmlir/detail/PR-779/11/pipeline

If everyone could please review so as to get this merged?

**Comment by zhanglx13** [comment]
What is the good convention to review such a large PR? github even cannot show files changed. Do we rely on the CI results?

**Comment by zhanglx13** [comment]
At least you can update the Future work section.
And maybe find a better place to keep track of the future works. Maybe open a ticket?

File: external/llvm-project/mlir/include/mlir/Dialect/Tosa/Transforms/Passes.h [hunk 1]
```diff
@@ -42,29 +42,8 @@ std::unique_ptr<Pass> createTosaMakeBroadcastablePass();
 std::unique_ptr<Pass> createTosaTestQuantUtilAPIPass();
 std::unique_ptr<Pass> createTosaOptionalDecompositions();
 std::unique_ptr<Pass> createTosaPartitionPass();
-
-class TosaPartitionPass : public TosaPartitionBase<TosaPartitionPass> {
```
Comment (sjw36) [inline external/llvm-project/mlir/include/mlir/Dialect/Tosa/Transforms/Passes.h:40]: @pcf000 should this remain for other users? Does it need a test.

Comment (krzysz00) [reply external/llvm-project/mlir/include/mlir/Dialect/Tosa/Transforms/Passes.h:40]: I've replaced all the tests with appropriate strings of options

**Comment by krzysz00** [comment]
To focus the reviewing load a bit:
- Nothing in a non-mlir directory is a reviewable change (that is, all the changes to clang, llvm, etc. have nothing to do with us)
- I've substantially touched, as best as I can find and recall:
  - The async dailect, mainly around `async.launch`
  - `external/llvm-project/mlir/lib/Dialect/Bufferization/Transforms/FuncBufferizableOpInterfaceImpl.cpp`
  - `external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaPartition.cpp` and, consequently, `external/llvm-project/mlir/test/lib/Dialect/Tosa/TosaTestPasses.cpp`
  - `mlir/include/mlir-c/RegisterRocMLIR.h`, `mlir/lib/CAPI/RegisterRocMLIR/*` and the `mlir/test/CAPI` directory
  - Moving `migraphx-transform` out of the conversion passes
  - mlir/include/mlir/Dialect/Rock/IR/RockOps.td
  - `mlir/lib/Conversion/TosaToRock/TosaToRock.cpp`
  - `mlir/lib/Dialect/Rock/Transforms/BufferizableOpInterfaceImpl.cpp`

**Review by sjw36 — APPROVED** [review]
LGTM.

**Pull request merged**